### PR TITLE
Document ShapeDtypeStruct

### LINKED ISSUE #15023: jax.ShapeDtypeStruct isn't documented
It is helpful for AOT compilation!

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #15023

### CODE REVIEW

File: jax/_src/api.py [hunk 1]
```diff
@@ -2704,6 +2704,16 @@ def device_get(x: Any):
...
 
 class ShapeDtypeStruct:
+  """A container for the shape, dtype, and other static attributes of an array.
+
+  ``ShapeDtypeStruct`` is often used in conjunction with :func:`jax.eval_shape`.
+
+  Args:
+    shape: a sequence of integers representing an array shape
+    dtype: a dtype-like object
+    naned_shape: (optional) a dictionary representing a named shape
```
Comment (sharadmv) [inline jax/_src/api.py:2714]: named_shape

**Review by sharadmv — APPROVED** [review]

File: jax/_src/api.py (hunk 1 above)
Comment (mattjj) [reply jax/_src/api.py:2714]: Can we not document `named_shape`? It's experimental and only partially implemented, and I'm not sure if it'll stick around forever.

Comment (mattjj) [reply jax/_src/api.py:2714]: In other words, I don't think it needs to be in the public API (though I may be mistaken).

Comment (jakevdp) [reply jax/_src/api.py]: Done, thanks!

**Pull request merged**